editor/src/components/editor/store/dispatch.tsx

```diff
@@ -742,6 +745,34 @@ function editorDispatchInner(
       }
     }
 
+    const actionNames = simpleStringifyActions(dispatchedActions)
+
+    // Check for duplicate UIDs that have originated from actions being applied.
+    const uniqueIDsResult = getAllUniqueUids(result.unpatchedEditor.projectContents)
+    if (uniqueIDsResult.duplicateIDs.length > 0) {
+      const errorMessage = `Running ${actionNames} resulted in duplicate UIDs ${JSON.stringify(
+        uniqueIDsResult.duplicateIDs,
+      )}.`
+      if (IS_TEST_ENVIRONMENT) {
+        // In tests blow out with an exception so that the error is correctly attributed.
+        throw new Error(errorMessage)
+      } else {
+        // When running in the browser log the error and tell the user to restart the editor.
+        console.error(errorMessage)
+        const errorToast = EditorActions.addToast(
+          notice(
+            `Utopia has suffered from an irrecoverable error, please reload the editor.`,
+            'ERROR',
+            true,
+          ),
+        )
+        result = {
+          ...result,
+          unpatchedEditor: UPDATE_FNS.ADD_TOAST(errorToast, result.unpatchedEditor),
+        }
```

Comment:
I think there're better ways to go about this. We should either keep only the toast, and write a pair of tests that assert that a toast is shown when duplicate uids are present, and that the toast is not shown otherwise. If we want to keep the exception, we should but this code into an object export and stub it with Sinon to throw an exception. I think interleaving production code with test code muddles up the implementation, and introduces a big side effect in a function that could otherwise be pure

Reply:
Previously the logic was always throwing the exception, which was not necessarily obvious from the call being made.

If we can stub it to throw an exception like that for both unit and browser tests, that sounds like it would be an improvement however.